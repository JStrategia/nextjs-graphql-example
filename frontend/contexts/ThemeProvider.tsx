import React, { createContext, memo, ReactNode, useMemo, useState } from 'react';
import { ThemeProvider as StyledComponentsThemeProvider } from 'styled-components';
import { ThemeMode } from '../types/styles';
import { getTheme, GlobalStyle } from '../styles';

const defaultMode = 'light';

type Props = {
  children: ReactNode;
};

function ThemeProvider({ children }: Props) {
  const [mode, setMode] = useState<ThemeMode>(defaultMode);

  const theme = useMemo(() => getTheme(mode), [mode]);

  function switchMode() {
    setMode(mode === 'light' ? 'dark' : 'light');
  }

  return (
    <ThemeContext.Provider value={{ mode, setMode, switchMode }}>
      <StyledComponentsThemeProvider theme={theme}>
        <GlobalStyle />
        {children}
      </StyledComponentsThemeProvider>
    </ThemeContext.Provider>
  );
}

const ThemeContext = createContext<{
  mode: ThemeMode;
  setMode: (mode: ThemeMode) => void;
  switchMode: () => void;
}>({ mode: defaultMode, setMode: () => void 0, switchMode: () => void 0 });

export default memo(ThemeProvider);

export { ThemeContext };
