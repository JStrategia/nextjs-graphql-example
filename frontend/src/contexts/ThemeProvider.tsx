import React, { createContext, ReactNode, useMemo, useState } from 'react';
import { ThemeProvider as StyledComponentsThemeProvider } from 'styled-components';
import { getTheme, GlobalStyle } from '../styles';
import { ThemeMode } from '../types/styles';

const defaultMode = 'dark';

type Props = {
  children: ReactNode;
  initialMode?: ThemeMode;
};

function ThemeProvider({ children, initialMode = defaultMode }: Props) {
  const [mode, setMode] = useState<ThemeMode>(initialMode);

  const theme = useMemo(() => getTheme(mode), [mode]);

  function switchMode() {
    setMode(mode === 'light' ? 'dark' : 'light');
  }

  return (
    <ThemeContext.Provider value={{ mode, setMode, switchMode }}>
      <StyledComponentsThemeProvider theme={theme}>
        <GlobalStyle />
        {children}
      </StyledComponentsThemeProvider>
    </ThemeContext.Provider>
  );
}

const ThemeContext = createContext<{
  mode: ThemeMode;
  setMode: (mode: ThemeMode) => void;
  switchMode: () => void;
}>({ mode: defaultMode, setMode: () => void 0, switchMode: () => void 0 });

export default ThemeProvider;
export { defaultMode, ThemeContext };
