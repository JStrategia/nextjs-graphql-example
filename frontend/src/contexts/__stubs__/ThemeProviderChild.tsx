import * as React from 'react';
import { useContext } from 'react';
import { ThemeContext } from 'contexts/ThemeProvider';
import { ThemeMode } from 'types/styles';

type Props = {
  readonly modeToSet: ThemeMode;
};

function ThemeProviderChild({ modeToSet }: Props) {
  const { mode, setMode, switchMode } = useContext(ThemeContext);

  return (
    <>
      <span data-testid="mode">{mode}</span>
      <button data-testid="switchModeButton" onClick={switchMode} />
      <button data-testid="setModeButton" onClick={() => setMode(modeToSet)} />
    </>
  );
}

export default ThemeProviderChild;
