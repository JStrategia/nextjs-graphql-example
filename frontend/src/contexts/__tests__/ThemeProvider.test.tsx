import * as React from 'react';
import { create } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import ThemeProvider from 'contexts/ThemeProvider';
import ThemeProviderChild from 'contexts/__stubs__/ThemeProviderChild';

describe('ThemeProvider', () => {
  describe('Given children', () => {
    const modeToSet = 'dark';
    const children = <ThemeProviderChild modeToSet={modeToSet} />;

    describe('When rendering', () => {
      const render = create(<ThemeProvider>{children}</ThemeProvider>).toJSON();

      test('Then match snapshot', () => {
        expect(render).toMatchSnapshot();
      });
    });

    describe('Given render', () => {
      const { getByTestId, rerender } = render(<ThemeProvider>{children}</ThemeProvider>);

      describe('Given no initial theme mode', () => {
        rerender(<ThemeProvider>{children}</ThemeProvider>);

        describe('When displaying theme mode', () => {
          const mode = getByTestId('mode');

          test('Then use dark mode', () => {
            expect(mode.innerHTML).toBe('dark');
          });
        });
      });

      describe('Given initial theme mode', () => {
        const initialMode = 'light';
        rerender(<ThemeProvider initialMode={initialMode}>{children}</ThemeProvider>);

        describe('When displaying theme mode', () => {
          const mode = getByTestId('mode');

          test('Then use initial theme mode', () => {
            expect(mode.innerHTML).toBe(initialMode);
          });
        });

        describe('When setting theme and displaying theme mode', () => {
          const setModeButton = getByTestId('setModeButton');
          fireEvent.click(setModeButton);
          const mode = getByTestId('mode');

          test('Then set theme mode', () => {
            expect(mode.innerHTML).toBe(modeToSet);
          });
        });

        describe('When switching theme and displaying theme mode', () => {
          const modeToBeSwitched = 'dark';
          const switchModeButton = getByTestId('switchModeButton');
          fireEvent.click(switchModeButton);
          const mode = getByTestId('mode');

          test('Then switch theme mode', () => {
            expect(mode.innerHTML).toBe(modeToBeSwitched);
          });
        });
      });
    });
  });
});
