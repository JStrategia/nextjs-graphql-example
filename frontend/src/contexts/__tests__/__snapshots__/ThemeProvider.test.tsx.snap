// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider Given children When rendering Then match snapshot 1`] = `
Array [
  <span
    data-testid="mode"
  >
    dark
  </span>,
  <button
    data-testid="switchModeButton"
    onClick={[Function]}
  />,
  <button
    data-testid="setModeButton"
    onClick={[Function]}
  />,
]
`;
